src/tools/illink/test/Mono.Linker.Tests.Cases/DataFlow/FeatureCheckDataFlow.cs

```diff
@@ -0,0 +1,999 @@
+			}
+
+			[ExpectedWarning ("IL3050", nameof (RequiresDynamicCode), ProducedBy = Tool.Analyzer | Tool.NativeAot)]
+			[ExpectedWarning ("IL3050", nameof (RequiresDynamicCode), ProducedBy = Tool.Analyzer | Tool.NativeAot)]
+			static void UnguardedTernary ()
+			{
+				var a = RuntimeFeature.IsDynamicCodeSupported ? true : RequiresDynamicCodeBool ();
+				var b = !RuntimeFeature.IsDynamicCodeSupported ? RequiresDynamicCodeBool () : true;
+			}
+
+			[ExpectedWarning ("IL3050", nameof (RequiresDynamicCode), ProducedBy = Tool.Analyzer | Tool.NativeAot)]
+			static void UnguardedThrow ()
+			{
+				if (RuntimeFeature.IsDynamicCodeSupported)
+				{
+					throw new Exception ();
+				}
+
+				RequiresDynamicCode ();
+			}
+
+			[ExpectedWarning ("IL3050", nameof (RequiresDynamicCode), ProducedBy = Tool.Analyzer | Tool.NativeAot)]
+			static void UnguardedAssert ()
+			{
+				Debug.Assert (!RuntimeFeature.IsDynamicCodeSupported);
+
+				RequiresDynamicCode ();
+			}
+
+			[ExpectedWarning ("IL3050", nameof (RequiresDynamicCode), ProducedBy = Tool.Analyzer | Tool.NativeAot)]
+			static void UnguardedDoesNotReturnIfTrue ()
+			{
+				DoesNotReturnIfTrue (RuntimeFeature.IsDynamicCodeSupported);
+
+				RequiresDynamicCode ();
+			}
+
+			[ExpectedWarning ("IL3050", nameof (RequiresDynamicCode), ProducedBy = Tool.Analyzer | Tool.NativeAot)]
+			static void UnguardedDoesNotReturnIfFalse ()
+			{
+				DoesNotReturnIfFalse (!RuntimeFeature.IsDynamicCodeSupported);
+
+				RequiresDynamicCode ();
+			}
+
+			[ExpectedWarning ("IL3050", nameof (RequiresDynamicCode), ProducedBy = Tool.Analyzer | Tool.NativeAot)]
+			static void UnguardedDoesNotReturn ()
+			{
+				if (RuntimeFeature.IsDynamicCodeSupported)
+					DoesNotReturn ();
+
+				RequiresDynamicCode ();
+			}
+
+			// NativeAot doesn't optimize branches away based on DoesNotReturnIfFalse
+			[ExpectedWarning ("IL3050", nameof (RequiresDynamicCode), ProducedBy = Tool.NativeAot)]
+			static void UnguardedDoesNotReturnIfFalseCtor ()
+			{
+				new DoesNotReturnIfFalseCtor (RuntimeFeature.IsDynamicCodeSupported);
+
+				RequiresDynamicCode ();
+			}
+
+			public static void Test ()
+			{
+				Unguarded ();
+				UnguardedIf ();
+				UnguardedElse ();
+				UnguardedAnd ();
+				UnguardedOr ();
+				UnguardedTernary ();
+				UnguardedThrow ();
+				UnguardedAssert ();
+				UnguardedDoesNotReturnIfTrue ();
+				UnguardedDoesNotReturnIfFalse ();
+				UnguardedDoesNotReturn ();
+				UnguardedDoesNotReturnIfFalseCtor ();
+			}
+		}
+
+		class CallFeatureGuarded
+		{
+			public static void Test ()
+			{
+				GuardedIf ();
+				GuardedElse ();
+				GuardedAnd ();
+				GuardedOr ();
+				GuardedTernary ();
+				GuardedThrow ();
+				GuardedAssert ();
+				GuardedDoesNotReturnIfTrue ();
+				GuardedDoesNotReturnIfFalse ();
+				GuardedDoesNotReturn ();
+				GuardedDoesNotReturnIfFalseCtor ();
+			}
+
+			static void GuardedIf ()
+			{
+				if (RuntimeFeature.IsDynamicCodeSupported)
+					RequiresDynamicCode ();
```

Comment:
Same as above - for a couple of these add RUC/RAF method calls to check that the feature check for RDC has no effect on those.